### PR TITLE
[ci] move Solaris and valgrind test steps into scripts

### PR DESCRIPTION
**Pull request by jameslamb**
#3946 introduces two new `Suggests`-level dependencies for the R package.

To run the Solaris and valgrind tests successfully, those dependencies need to be added to the jobs for those tests (https://github.com/microsoft/LightGBM/blob/master/.github/workflows/r_solaris.yml and https://github.com/microsoft/LightGBM/blob/master/.github/workflows/r_valgrind.yml).

This creates an interesting challenge. Right now changes like that need to be made in the workflow files in `.github/workflows/`, but when those jobs are triggered by comments they source those configs from `master` (https://github.com/microsoft/LightGBM/pull/4397#issuecomment-866790489).

This PR proposes moving more of the logic for those checks into scripts in `.ci/`. That will make it possible to test the following types of changes to the Solaris and valgrind jobs before merging them:

* adding or removing dependencies installed with `install.packages()`
* changing arguments to `install.packages()` (e.g. #4198)
* changing `apt-get install`'d dependencies for the Solaris checks
* changing arguments passed to `Rdvalgrind CMD INSTALL` in the valgrind checks

### CODE REVIEW

File: .ci/test_r_package_solaris.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+sh build-cran-package.sh || exit -1
+
+apt-get install --no-install-recommends -y \
+  libcurl4-openssl-dev \
+  libxml2-dev \
+  libssl-dev
+
+log_file="$GITHUB_WORKSPACE/rhub_logs.txt"
```
Comment (StrikerRUS) [inline .ci/test_r_package_solaris.sh:10]: We try to not use CI service dependable env variables in bash script (try to search for `GITHUB_WORKSPACE`: it is used only in config `yml` files).

I guess we can switch to relative paths here just like in `test_r_package_valgrind.sh`
https://github.com/microsoft/LightGBM/blob/2f3764265cc2c43ffaa250e89c1f2efc859404ae/.ci/test_r_package_valgrind.sh#L5-L13

File: .github/workflows/r_solaris.yml [hunk 2]
```diff
@@ -37,15 +37,7 @@ jobs:
...
-        run: |
-          sh build-cran-package.sh || exit -1
-          apt-get install --no-install-recommends -y \
-              libcurl4-openssl-dev \
-              libxml2-dev \
-              libssl-dev
-          log_file="$GITHUB_WORKSPACE/rhub_logs.txt"
-          Rscript -e "install.packages('rhub', dependencies = c('Depends', 'Imports', 'LinkingTo'), repos = 'https://cran.r-project.org', Ncpus = parallel::detectCores())"
-          Rscript $GITHUB_WORKSPACE/.ci/run_rhub_solaris_checks.R $(pwd)/lightgbm_*.tar.gz $log_file || exit -1
+        run: ./.ci/test_r_package_valgrind.sh
```
Comment (StrikerRUS) [inline .github/workflows/r_solaris.yml:40]:
```suggestion
        run: ./.ci/test_r_package_solaris.sh
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Really useful refactoring! Thanks!

Please check my initial comments below.

File: .ci/test_r_package_solaris.sh (hunk 1 above)
Comment (jameslamb) [reply .ci/test_r_package_solaris.sh]: good point! fixed in https://github.com/microsoft/LightGBM/pull/4503/commits/10deff608ebcd95b60abd310472bad181d9768e2

**Review by StrikerRUS — APPROVED** [review]
LGTM! Let's check these changes in a next PR.

**Pull request merged**